lib/network_connection.go

```diff
@@ -0,0 +1,212 @@
+// OutboundConnectionAttempt concludes. outboundConnection implements the Connection interface.
+type outboundConnection struct {
+	mtx        sync.Mutex
+	terminated bool
+
+	attemptId    uint64
+	address      *wire.NetAddress
+	connection   net.Conn
+	isPersistent bool
+	failed       bool
+}
+
+func (oc *outboundConnection) GetConnectionType() ConnectionType {
+	return ConnectionTypeOutbound
+}
+
+func (oc *outboundConnection) Close() {
+	oc.mtx.Lock()
+	defer oc.mtx.Unlock()
+
+	if oc.terminated {
+		return
+	}
+	oc.connection.Close()
+	oc.terminated = true
+}
+
+// inboundConnection is used to store an established connection with a peer. inboundConnection is created after a
+// an external peer connects to the node. inboundConnection implements the Connection interface.
+type inboundConnection struct {
+	mtx        sync.Mutex
+	terminated bool
+
+	connection net.Conn
+}
+
+func (ic *inboundConnection) GetConnectionType() ConnectionType {
+	return ConnectionTypeInbound
+}
+
+func (ic *inboundConnection) Close() {
+	ic.mtx.Lock()
+	defer ic.mtx.Unlock()
+
+	if ic.terminated {
+		return
+	}
+
+	ic.connection.Close()
+	ic.terminated = true
+}
+
+// OutboundConnectionAttempt is used to store the state of an outbound connection attempt. It is used to initiate
+// an outbound connection to a peer, and manage the lifecycle of the connection attempt.
+type OutboundConnectionAttempt struct {
+	mtx sync.Mutex
+
+	// attemptId is used to identify the connection attempt. It will later be the id of the peer,
+	// if the connection is successful.
+	attemptId uint64
+
+	// netAddr is the address of the peer we are attempting to connect to.
+	netAddr *wire.NetAddress
+	// isPersistent is used to indicate whether we should retry connecting to the peer if the connection attempt fails.
+	// If isPersistent is true, we will retry connecting to the peer until we are successful. Each time such connection
+	// fails, we will sleep according to exponential backoff. Otherwise, we will only attempt to connect to the peer once.
+	isPersistent bool
+	// dialTimeout is the amount of time we will wait before timing out an individual connection attempt.
+	dialTimeout time.Duration
+	// timeoutUnit is the unit of time we will use to calculate the exponential backoff delay. The initial timeout is
+	// calculated as timeoutUnit * 2^0, the second timeout is calculated as timeoutUnit * 2^1, and so on.
+	timeoutUnit time.Duration
+	// retryCount is the number of times we have attempted to connect to the peer.
+	retryCount uint64
+	// connectionChan is used to send the result of the connection attempt to the caller thread.
+	connectionChan chan *outboundConnection
+
+	exitChan chan bool
+	status   outboundConnectionAttemptStatus
+}
+
+type outboundConnectionAttemptStatus int
+
+const (
+	outboundConnectionAttemptUnknown    outboundConnectionAttemptStatus = 0
+	outboundConnectionAttemptStarted    outboundConnectionAttemptStatus = 1
```

Comment:
nit: these are more standard const names that match the terminology we use elsewhere
```suggestion
	outboundConnectionAttemptInitialized    outboundConnectionAttemptStatus = 0
	outboundConnectionAttemptRunning      outboundConnectionAttemptStatus = 1
```